packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -73,13 +91,21 @@ export class ConnectionStateHandler {
         // timing out ops fetches. So attempt recovery infrequently. Also fetch uses 30 second timeout, so
         // if retrying fixes the problem, we should not see these events.
         this.joinOpTimer = new Timer(
-            JoinOpTimer,
+            JoinOpTimeoutMs,
             () => {
                 // I've observed timer firing within couple ms from disconnect event, looks like
                 // queued timer callback is not cancelled if timer is cancelled while callback sits in the queue.
-                if (this.connectionState === ConnectionState.Connecting) {
-                    this.handler.logConnectionIssue("NoJoinOp");
+                if (this.connectionState !== ConnectionState.Connecting) {
+                    return;
                 }
+                const quorumClients = this.handler.quorumClients();
+                const details = {
+                    quorumInitialized: quorumClients !== undefined,
+                    hasPendingClientId: this.pendingClientId !== undefined,
+                    inQuorum: quorumClients?.getMember(this.pendingClientId ?? "") !== undefined,
+                    waitingForLeaveOp: this.waitingForLeaveOp,
+                };
+                this.handler.logConnectionIssue("NoJoinOp", details);
```

Comment:
@jatgarg @vladsud - There is a theoretical case where we would never transition to Connected state:

1. "connect" event fires before quorum is initialized
2. quorum initializes with the just-connected client provided during construction
3. Now the "pending" client is in the quorum, but no "addMember" event fired.
4. We'll hit this timeout, and will never connect.

I don't imagine this is possible today - the only place doing (2) is something about a snapshot and I doubt that would include the just-established connection in a race condition.  But it's a possibility given this state machine, so I think I'll open a bug on it - the fix could be to add an assert to fail fast, or actually try to support this flow.

Reply:
I'm not sure I follow, specifically # 3. I assume you talk about "write" connection.
After protocol finishes initialization (assuming it's boot sequence), we allow ops processing to start and will process join op. And that will transition container to connected state.

Can you please clarify if you talk about reconnect or first connection, and it's relationship to boot sequence (i.e. are we talking about connection happening at the very top of Container.load(), or after load() is done).

Reply:
For posterity:  This case isn't hit in production in last 30 days.  All "NoJoinOp" events have **inQuorum: false** as you'd expect (but in the corner case it would be true).